src/core/api.c

```diff
@@ -1325,20 +1325,32 @@ MsQuicStreamReceiveComplete(
 
     QUIC_CONN_VERIFY(Connection,
         (Stream->RecvPendingLength == 0) || // Stream might have been shutdown already
-        BufferLength <= Stream->RecvPendingLength);
+        BufferLength <= Stream->RecvPendingLength ||
+        BufferLength <= (UINT64_MAX >> 2));
 
     QuicTraceEvent(
         StreamAppReceiveCompleteCall,
         "[strm][%p] Receive complete call [%llu bytes]",
         Stream,
         BufferLength);
 
-    InterlockedExchangeAdd64(
-        (int64_t*)&Stream->RecvCompletionLength, (int64_t)BufferLength);
+    uint64_t RecvCompletionLength = InterlockedExchangeAdd64(
+        (int64_t*)&Stream->RecvCompletionLength,
+        (int64_t)BufferLength);
+    if ((BufferLength & QUIC_STREAM_RECV_COMPLETION_LENGTH_CANARY_BIT) != 0 &&
+        (RecvCompletionLength & QUIC_STREAM_RECV_COMPLETION_LENGTH_CANARY_BIT) != 0) {
+        // overflow detected
+        QuicTraceEvent(
+            ConnError,
+            "[conn][%p] ERROR, %s.",
+            Connection,
+            "RecvCompletionLength is overflow");
+        QuicConnSilentlyAbort(Connection);
```

Comment:
I think we should just return an error, not abort.

Reply:
I mean don't call the abort function. Just return invalid_state or something.

Reply:
@nibanks `MsQuicStreamReceiveComplete` returns `void`, so returning an error code is not an option.
IMO, if `RecvCompletionLength` overflows, it is a clear sign of an application error (no way we could ever complete UINT64_MAX / 2 bytes at once).

- Either we treat it as an irrecoverable error (we can't know how much the app intended to complete, so we can't deliver the next bytes reliably and in order) -> in that case, we either abort the connection or end the process.
- We could treat it as "complete everything pending", but I think it would only be confusing and let a bad state persist longer

I don't think that logging a warning and returning without action is a good solution, in that case, we should simply assume the overflow can't happen and not use a canary bit.